### PR TITLE
remove let-env, focus on mutating $env

### PR DESCRIPTION
**Pull request by jntrnr**
# Description

For years, Nushell has used `let-env` to set a single environment variable. As our work on scoping continued, we refined what it meant for a variable to be in scope using `let` but never updated how `let-env` would work. Instead, `let-env` confusingly created mutations to the command's copy of `$env`.

So, to help fix the mental model and point people to the right way of thinking about what changing the environment means, this PR removes `let-env` to encourage people to think of it as updating the command's environment variable via mutation.

Before:

```
let-env FOO = "BAR"
```

Now:

```
$env.FOO = "BAR"
```

It's also a good reminder that the environment owned by the command is in the `$env` variable rather than global like it is in other shells.

# User-Facing Changes

BREAKING CHANGE BREAKING CHANGE

This completely removes `let-env FOO = "BAR"` so that we can focus on `$env.FOO = "BAR"`.

# Tests + Formatting

# After / Before Submitting
integration scripts to update:
- :heavy_check_mark: [starship](https://github.com/starship/starship/blob/master/src/init/starship.nu)
- :heavy_check_mark: [virtualenv](https://github.com/pypa/virtualenv/blob/main/src/virtualenv/activation/nushell/activate.nu)
- :heavy_check_mark: [atuin](https://github.com/ellie/atuin/blob/main/atuin/src/shell/atuin.nu) (PR: https://github.com/ellie/atuin/pull/1080)
- :x: [zoxide](https://github.com/ajeetdsouza/zoxide/blob/main/templates/nushell.txt) (PR: https://github.com/ajeetdsouza/zoxide/pull/587)
- :heavy_check_mark: [oh-my-posh](https://github.com/JanDeDobbeleer/oh-my-posh/blob/main/src/shell/scripts/omp.nu) (pr: https://github.com/JanDeDobbeleer/oh-my-posh/pull/4011)

### CODE REVIEW

**Review by amtoine — COMMENTED** [review]
for now it looks good to me :ok_hand: 

of course this breaks integration with other tools :+1:

**Comment by amtoine** [comment]
@jntrnr, the failing test is [here](https://github.com/nushell/nushell/pull/9574/files#diff-a14193f0d465e8a19db9f52fe852099e0733dd0b7147be5445173ea825c272cc), looks like the behaviour of `$env.FUIQW = ...` has changed :thinking:

**Comment by jntrnr** [comment]
Looks like starship and virtual already use `load-env` and aren't affected.

**Comment by jntrnr** [comment]
atuin looks like it will break: https://github.com/ellie/atuin/blob/main/atuin/src/shell/atuin.nu

**Review by amtoine — COMMENTED** [review]
i do not understand why we need to use `load-env` in the last commit :confused:

**Comment by jntrnr** [comment]
@amtoine because the path variable is coming from a variable. `$env.$FOO = ...` isn't supported, only `$env.FOO = ...`

**Comment by amtoine** [comment]
@jntrnr, i've added the integration scripts above to the body :ok_hand:

**Comment by amtoine** [comment]
> @amtoine because the path variable is coming from a variable. `$env.$FOO = ...` isn't supported, only `$env.FOO = ...`
>
oooooh very good catch :scream: :clap: :clap: 
i did not even see the `$` :eyes:

**Comment by jntrnr** [comment]
Zoxide looks like it will also break: https://github.com/ajeetdsouza/zoxide/blob/main/templates/nushell.txt

**Comment by amtoine** [comment]
i propose to make this a draft to avoid merging the PR before we have fixed the integration scripts :open_mouth:

**Review by amtoine — CHANGES_REQUESTED** [review]
maybe `LetEnvDeprecated` should go inside `crates/nu-command/src/deprecated/` as the other deprecated commands? :thinking:

**Review by amtoine — APPROVED** [review]
looks about right to me... LET'S BREAK IT :muscle: :triumph:

File: crates/nu-command/src/deprecated/let_env.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    fn run(
+        &self,
+        _: &EngineState,
+        _: &mut Stack,
+        call: &Call,
+        _: PipelineData,
+    ) -> Result<PipelineData, ShellError> {
+        Err(nu_protocol::ShellError::DeprecatedCommand(
+            self.name().to_string(),
+            "$nu.<environment variable> = ...".to_owned(),
```
Comment (MangoIV) [inline crates/nu-command/src/deprecated/let_env.rs:39]: I think this should be `$env`

Comment (fdncred) [reply crates/nu-command/src/deprecated/let_env.rs:39]: good catch. thanks

**Comment by hyiltiz** [comment]
This is a great change, thank you! I was getting quite confused, but this also breaks quite some scripts. Since the fix is mechanical, would it be meaningful to produce `script.nu.orig -> script.nu` by applying the `let-env FOO = ...` -> `$env.FOO = ...` conversion automatically to help ease the BREAKING-CHANGE transition. Similar procedure can be used for other breaking changes as well.

**Comment by amtoine** [comment]
> This is a great change, thank you! I was getting quite confused, but this also breaks quite some scripts. Since the fix is mechanical, would it be meaningful to produce `script.nu.orig -> script.nu` by applying the `let-env FOO = ...` -> `$env.FOO = ...` conversion automatically to help ease the BREAKING-CHANGE transition. Similar procedure can be used for other breaking changes as well.
>
a few commands i know that should do this migration :yum: 
- `sed -i 's/let-env /$env./' ...
- `sd 'let-env ' '$env.' ...
- `open ... | str replace --all --string 'let-env ' '$env.' | save --force ...`

i did not try them but i think they should do it :yum:

**Comment by hyiltiz** [comment]
I had an alias `alias "go home" = let-env JAVA_HOME = (/usr/libexec/java_home -a x86_64 -v 11)` before and after transition, `alias "go home" = $env.JAVA_HOME = (/usr/libexec/java_home -a x86_64 -v 11)`, I get the helpful message that `aliasing BinaryOp is not supported`. I really hope aliases are agnostic to the content of what they are being alieased to. I can probably work around this by defining `go home` as a function instead, and maybe with the `load-env` trick, but this probably should be possible in a more straightforward way.

**Comment by amtoine** [comment]
@hyiltiz 
yeah, aliases should be command calls, this is why `$env.PY = ...` does not work. and `do { ... }` does not work either because i'd say `do` is a bit like a keyword?

`load-env` is a great solution :ok_hand: 
you can also write a custom `def-env` command for such things :yum:

**Comment by hyiltiz** [comment]
Is there a benefit for defining it as a command vs. alias, except the fact that in a command there is much larger flexibility to do arbitrarily complex things (may not be desirable since by alias I meant to show the shallowness of implementation).

I'll just use `load-env` and maybe convert them to `def-env` if needs arise. Thanks, @amtoine!

**Comment by amtoine** [comment]
> Is there a benefit for defining it as a command vs. alias, except the fact that in a command there is much larger flexibility to do arbitrarily complex things (may not be desirable since by alias I meant to show the shallowness of implementation).
> 
maybe @kubouch might have more information about the aliases as he rewrote them completely not so long ago, but i'd say that by design aliases do alias to commands, not language constructs.
e.g. you can `alias ll = ls -l` because `ls -l` is a *simple* command call but you cannot `alias foo = if ...` because this is not a command :thinking: 

> I'll just use `load-env` and maybe convert them to `def-env` if needs arise. Thanks, @amtoine!
>
yeah, as `load-env` is a command, it does the trick :smirk:

**Comment by EnricoGhiorzi** [comment]
This did break compatibility with some other tools (I experience it with Starship) because the updated scripts have not been included into a release yet (not to mention it required hand-fixing the env and config files). I guess it's too late now, but instead of turning the old sytnax into a hard error it would have been better to deprecate it (as in, just rise a warning) for some time. Hopefully this will fix itself as the new scripts are released. Overall, not a great user experience.

**Comment by amtoine** [comment]
> This did break compatibility with some other tools (I experience it with Starship) because the updated scripts have not been included into a release yet (not to mention it required hand-fixing the env and config files).
>
yeah, the fix i wrote on the Starship side is ahead the latest release by 7 commits :sob: 

> I guess it's too late now, but instead of turning the old sytnax into a hard error it would have been better to deprecate it (as in, just rise a warning) for some time.
>
the deprecation system definitely needs love :relieved: 
we will work on this in https://github.com/nushell/nushell/issues/9812!

> Hopefully this will fix itself as the new scripts are released.
>
i can confirm you it will :ok_hand: 
i build Starship from source and it works like a charm with latest release!

> Overall, not a great user experience.
>
again, sorry for the inconvenience...
we are working on release a patch as soon as possible :crossed_fingers:

**Pull request merged**